client/src/main/java/io/pravega/client/netty/impl/SessionHandler.java

```diff
@@ -0,0 +1,235 @@
+    private final AtomicReference<Channel> channel = new AtomicReference<>();
+    private final AtomicReference<ScheduledFuture<?>> keepAliveFuture = new AtomicReference<>();
+    private final AtomicBoolean recentMessage = new AtomicBoolean(false);
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final AppendBatchSizeTracker batchSizeTracker;
+    private final ReusableFutureLatch<Void> registeredFutureLatch = new ReusableFutureLatch<>();
+    private final ConcurrentHashMap<Integer, ReplyProcessor> sessionIdReplyProcessorMap = new ConcurrentHashMap<>();
+    private final AtomicBoolean disableSession = new AtomicBoolean(false);
+
+    public SessionHandler(String connectionName, AppendBatchSizeTracker batchSizeTracker) {
+        this.connectionName = connectionName;
+        this.batchSizeTracker = batchSizeTracker;
+    }
+
+    /**
+     * Create a session on existing connection.
+     * @param session Session.
+     * @param rp ReplyProcessor for the specified session.
+     * @return Client Connection object.
+     */
+    public ClientConnection createSession(final Session session, final ReplyProcessor rp) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Preconditions.checkState(!disableSession.get(), "Ensure sessions are enabled");
+        log.info("Creating Session: {} for Endpoint: {}. The current Channel is {}.", session.getSessionId(), connectionName,
+                 channel.get());
+        if (sessionIdReplyProcessorMap.put(session.getSessionId(), rp) != null) {
+            throw new IllegalArgumentException("Multiple sessions cannot be created with the same Session id {}" + session.getSessionId());
+        }
+        return new ClientConnectionImpl(connectionName, session.getSessionId(), batchSizeTracker, this);
+    }
+
+    /**
+     * Create a {@link ClientConnection} where sessions are disabled. This implies that there is only one session on the underlying
+     * network connection.
+     * @param rp  The ReplyProcessor.
+     * @return Client Connection object.
+     */
+    public ClientConnection createConnectionWithSessionDisabled(final ReplyProcessor rp) {
+        Exceptions.checkNotClosed(closed.get(), this);
+        Preconditions.checkState(!disableSession.getAndSet(true), "Sessions are disabled, incorrect usage patter");
+        log.info("Creating a new connection with session disabled for Endpoint: {}. The current Channel is {}.", connectionName,
+                 channel.get());
+        sessionIdReplyProcessorMap.put(SESSION_DISABLED, rp);
+        return new ClientConnectionImpl(connectionName, SESSION_DISABLED, batchSizeTracker, this);
+    }
+
+    /**
+     * Close a session. This is invoked when the ClientConnection is closed.
+     * @param clientConnection Client Connection.
+     */
+    public void closeSession(ClientConnection clientConnection) {
+        int session = ((ClientConnectionImpl) clientConnection).getSession();
```

Comment:
Put this cast connection into its own variable instead of casting twice.

Reply:
fixed it.